File: algorithms/strings/game-of-thrones/game-of-thrones-javascript/js/gameOfThrones.js
// # A tarefa do problema "Game of Thrones - I" 
// # no HackerRank é determinar se é possível 
// # reorganizar as letras de uma string para 
// # formar um palíndromo. Um palíndromo é uma 
// # palavra que permanece a mesma se lida de 
// # trás para frente. Se a reorganização das 
// # letras da string puder formar um palíndromo, 
// # o programa deve imprimir "YES", caso contrário,
// # deve imprimir "NO".

// Exporta a função gameOfThrones como uma função nomeada
export function gameOfThrones(s) {
    
    // Converte a string s em um array de caracteres e ordena em ordem alfabética
    const charArr = [...s].sort();

    // Cria um objeto para contar as ocorrências de cada caractere
    const contarCaracter = {}

    // Itera sobre cada caractere no array charArr
    charArr.forEach(c => {

        // Verifica se o caractere já existe no objeto contarCaracter
        if (contarCaracter[c]) {
            
            // Se sim, incrementa a contagem de ocorrências desse caractere
            contarCaracter[c]++;

        } else {

            // Se não, adiciona o caractere ao objeto com uma contagem de 1
            contarCaracter[c] = 1;
        }
    });

    // Inicializa um contador para contar o número de caracteres com ocorrências ímpares
    let contarImpar = 0;

    // Itera sobre os valores do objeto contarCaracter
    for(const contador of Object.values(contarCaracter)) {

        // Verifica se o contador é ímpar
        if (contador % 2 !== 0) {
            
            // Se sim, incrementa o contador de caracteres com ocorrências ímpares
            contarImpar++;

        }

        // Se houver mais de um caractere com ocorrências ímpares, 
        // não é possível formar um palíndromo
        if (contarImpar > 1) {
            
            // Retorna "NO"
            return "NO";
        }
    }

    // Se não houver mais de um caractere com ocorrências ímpares, é possível formar um palíndromo
    // Retorna "YES"
    return "YES";
}

File: algorithms/strings/game-of-thrones/game-of-thrones-python/tests/gameOfThrones.py
# A tarefa do problema "Game of Thrones - I" 
# no HackerRank é determinar se é possível 
# reorganizar as letras de uma string para 
# formar um palíndromo. Um palíndromo é uma 
# palavra que permanece a mesma se lida de 
# trás para frente. Se a reorganização das 
# letras da string puder formar um palíndromo, 
# o programa deve imprimir "YES", caso contrário,
# deve imprimir "NO".

def gameOfThrones(s):
    
    # Converte a string em uma lista de caracteres
    lista_s = list(s)
    
    # Inicializa um contador
    contador = 0
    
    # Itera sobre o conjunto de caracteres únicos na string
    for i in set(s):
        
        # Verifica se a contagem de ocorrências do caractere na lista é ímpar
        if lista_s.count(i) % 2 != 0:
            
            # Se for ímpar, incrementa o contador
            contador = contador + 1
            
        # Se o contador exceder 1, significa que mais de um caractere tem um número ímpar de ocorrências,
        # o que torna impossível formar um palíndromo
        if contador > 1:
            
            # Retorna "NO"
            return "NO"
    
    # Se não houver mais de um caractere com ocorrências ímpares, é possível formar um palíndromo
    # Retorna "YES"
    return "YES"